Added so equals and hashcode is generated in models.

denvitaharen opened the pull request:
Many thanks for submitting your Pull Request :heart:!

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors guide](CONTRIBUTING.md)
- [x] Your code is properly formatted according to [our code style](CONTRIBUTING.md#code-style)
- [x] Pull Request title contains the target branch if not targeting main: `[0.9.x] Subject`
- [ ] Pull Request contains link to the issue
- [ ] Pull Request contains link to any dependent or related Pull Request
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket

I have added so `equals` and `hashCode` is generated to the models. It's possible to disabled it if someone doesn't want it.

@ricardozanini I created an new branch and pull request, it was easier and faster for me to do that.


The result looks like this:

```
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        Animal model = (Animal) o;

        return Objects.equals(born, model.born) &&
        Objects.equals(deceased, model.deceased);
    }

    @Override
    public int hashCode() {
        return Objects.hash(born,
        deceased);
    }
```

ricardozanini commented on docs/modules/ROOT/pages/includes/equals-hashcode.adoc:
```diff
@@ -0,0 +1,9 @@
+`Hashcode` and `equals` are by default generated. It adds all the variables in the model
+to the model to the methods.
```
```suggestion
`Hashcode` and `equals` are by default generated. It adds all the variables in the model to the model to the methods.
```
I am not sure if I understood this sentence. 🤔

denvitaharen replied:
In my head it was 100% clear when I pushed it :D I will try and refrase it! :)

denvitaharen replied:
I rewrote the documentation, I hope this ones better.

denvitaharen commented:
```
    /**
     * Compares this object to the specified object. The result is
     * {@code true} if and only if the argument is not
     * {@code null} and is a {@code Animal} object that
     * contains the same values as this object.
     *
     * @param   obj   the object to compare with.
     * @return  {@code true} if the objects are the same;
     *          {@code false} otherwise.
     **/
    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;

        Animal model = (Animal) obj;

        return Objects.equals(born, model.born) &&
        Objects.equals(deceased, model.deceased);
    }

    /**
     * Returns a hash code for a {@code Animal}.
     *
     * @return a hash code value for a {@code Animal}.
     **/
    @Override
    public int hashCode() {
        return Objects.hash(born,
        deceased);
    }
```

I realized that I hadn't written any javadoc for the new methods, so I added it now. It took some inspiration from the String and Long javadoc and rewrote so it fits this in.

mbard commented:
Excellent, I really need equal/hashcode

ricardozanini reviewed (approved):
LGTM!

hbelmiro commented on client/integration-tests/equals-hashcode/src/test/java/io/quarkiverse/openapi/generator/it/VerifyGenerationTest.java:
```diff
@@ -0,0 +1,47 @@
+
+    private static Stream<Arguments> provideStringsForIsBlank() {
+        return Stream.of(
+                Arguments.of(Animal.class, "equals", Object.class),
+                Arguments.of(Animal.class, "hashCode", Object.class),
```
Do we really need to check `Object`?

denvitaharen replied:
I thought a long time of how to verify that we don't generate `equals`and ´hashCode` and the best way I thought out was to get what class that declares a method. This way I can check if the generated model has or does not have the methods based on what set in the `application.properties`.

In the case of the models that doesn't have a generated `equals` and `hashCode`, the methods are declared in ´Object´.

What I can do is I can split the tests and don't check against `Object` and instead just check that the methods its not declared in the generated model. Or I can leave it like this?

Or does somebody have an better way of verifying the generation?

hbelmiro replied:
My point is that `Object` will always have `equals` and `hashCode` declared. The verification will always pass.

denvitaharen replied:
Thats a valid point @hbelmiro, I now see the problem with my test. I have rewritten the test so it only tests the models instead.

hbelmiro commented on client/deployment/src/main/resources/templates/libraries/microprofile/pojo.qute:
```diff
@@ -1,5 +1,6 @@
 {@org.openapitools.codegen.CodegenModel m}
 import com.fasterxml.jackson.annotation.JsonProperty;
+import java.util.Objects;
```
Instead of importing, can you please use the FQCN to avoid compilation errors due to possible ambiguities?

denvitaharen replied:
> Instead of importing, can you please use the FQCN to avoid compilation errors due to possible ambiguities?

Ofc, I will change so it uses FQCN instead. I tried to match the existing patterns and StringBuilder doesn't have use FQCN.

hbelmiro replied:
Just to complement, we still have some imports, like `com.fasterxml.jackson.annotation.JsonProperty` in the above line. But the idea is to remove them and not add more.

mcruzdev commented on docs/modules/ROOT/pages/includes/equals-hashcode.adoc:
```diff
@@ -0,0 +1,9 @@
+`hashcode` and `equals` are by default generated to all models. `hashcode` and `equals` is based on all the variables
+in the model and can't be customized.
```
```suggestion
By default, `hashcode` and `equals` methods are automatically generated for all models. These methods are based on all the variables in the model and can not be customized.
```

mcruzdev replied:
Nice PR!

hbelmiro commented on client/integration-tests/equals-hashcode/src/test/java/io/quarkiverse/openapi/generator/it/VerifyGenerationTest.java:
```diff
@@ -1,42 +1,37 @@
+        var hashCodeMethod = getMethod(org.acme.non.equals.hashcode.model.Animal.class, "hashCode");
+
+        assertNotEquals(equalsMethod.getDeclaringClass(), org.acme.equals.hashcode.model.Animal.class);
+        assertNotEquals(hashCodeMethod.getDeclaringClass(), org.acme.equals.hashcode.model.Animal.class);
```
I think the method names are inverted.
